awx_collection/tools/templates/tower_module.j2

```diff
@@ -141,10 +163,17 @@ def main():
 {% if item['json']['actions']['POST'][option]['type'] == 'id' %}
     {{ option }}_id = None
     if {{ option }}:
-        {{ option }}_id = module.resolve_name_to_id('{{ option }}', {{ option }})
+        {{ option }}_id = module.resolve_name_to_id('{{ name_to_id_endpoint_resolution[option] }}', {{ option }})
```

Comment:
won't `name_to_id_endpoint_resolution[option]` give a key error for most types?

Reply:
Yes, we will need to keep populating the resolution.yml file out as we 'generate' modules. This started when I was looking at tower_job_template and got an association for webhook_credential. There is no webhook_credential end point so I need a mapping to know that webhook_credential is looked up via credentials.

Reply:
I gave this serious thought, here's what I think right now:

This is addressing a problem in the _code generator_. However, this particular problem is much better solved in _code_.

There is a "normal" case, where the ForeignKey name can be pluralized with a simple `"s"`. Related object `project` is found from the `projects` list. Maintaining this YAML translation file means that every case has to be handled.

The module utils class can have a method that performs this function "you give me a related field name, and I'll give you the global endpoint for the resource". If we put it there, we can unit test it as it expands. If the entire model breaks down because field names are repeated in different cases, but reference different model types, then we can add optional keyword args to the method. That just puts us in a much better place to maintain this.

Reply:
Pluralization is the easy part, its the translating of things like webhook_credential into credentials but we could certainly have a map inside the function. 

I love the idea of having a function and being able to test it but would we be able to pull all of the various verbs and their end points programmatically for the test? If so, I'm all in on moving this to module utils and dropping the generator lookup.

If we can't get all of the verbs and endpoints for the test, I think having this map in the generator is actually better because you could catch something thats missing at "generation" time of the module rather than "run time". i.e. if the hash can't resolve something the ansible generate fails, you fix the hash, ponder if the end point is deviating for a good reason and then generate again. VS. if its in module_utils and something slips by you won't know it until the module is running in a test (and if we aren't testing that particular parameter it could slip through the cracks down to the end user.

Reply:
If something is in the generator, as opposed to the code, that makes it much harder for people to track down. We are expected to maintain the code, not the generator. For that reason, I think `param_to_endpoint` makes the most sense. I use the word "param" where you are using "option" here most of the time. But I'm trying not to dwell on it too much, because I don't want to keep the overall paradigm of name-to-id at all. It runs against trends in the server development.